### PR TITLE
Allow for user groups.

### LINKED ISSUE #48: Add access control mechanism to geoDB
The geoDB needs access control, as in unix (discriminate between owner, group, other). The driving use case is to allow projects to work with different people.

### PR DESCRIPTION
**Pull request by thomasstorm**
This PR closes #48. A new set of functions have been added to the client, allowing to create user groups, to add/remove users to/from groups, to publish/unpublish collections to/from groups, and receive information about existing groups.

Checklist:

* [x] Issue has been created for change
* [x] Added unit tests and/or doctests in docstrings
* [x] Added docstrings and API docs for any new/modified user-facing classes and functions
* [x] New/modified features documented in `docs/source/*`
* [x] Changes documented in `CHANGES.md`
* [x] GitHub CI passes
* [x] Test coverage remains or increases (target 100%)

### CODE REVIEW

**Comment by thomasstorm** [comment]
@forman @AliceBalfanz: FYI in case you'd like to review. I have added a new notebook demonstrating how the group functions work; that could be a starting point for testing. Thanks!

**Review by forman — APPROVED** [review]
Looks great! You should add API docs.

**Comment by thomasstorm** [comment]
> Looks great! You should add API docs.

Thanks :) "add API docs" - what do you mean? To the new API (in `geodb.py`), I did add docs.

File: tests/core/test_message.py [hunk 1]
```diff
@@ -7,7 +7,3 @@ class TestMessage(unittest.TestCase):
     def test_message(self):
         message = Message('helge ist durchaus klug')
```
Comment (AliceBalfanz) [inline tests/core/test_message.py:8]:
```suggestion
        message = Message('Thomas ist durchaus klug')
```

**Review by AliceBalfanz — COMMENTED** [review]
I need some more clarifications: 
1. when I create a user group, am I automatically the group admin?
2. can every user create a user group?
3. how do I restrict access to a usergroup to read only? Based on the output:
`geodb.get_access_rights(collection_name)` 
it looks like the other members may modify the collection:
```
{'geodb_139b9f28-6757-4ffa-8345-3f658eb33850': ['INSERT',
  'SELECT',
  'UPDATE',
  'DELETE',
  'TRUNCATE',
  'REFERENCES',
  'TRIGGER'],
 'geodb_139b9f28-6757-4ffa-8345-3f658eb33850_test_group': ['INSERT',
  'SELECT',
  'UPDATE',
  'DELETE',
  'TRUNCATE',
  'REFERENCES',
  'TRIGGER']}
```
4. can I delete a user-group? 
5. in the notebook: 

> Create a new group (ensure that it does not already exist)  

shouldn't it be ensured on the client side, that I cannot create a user group with the same name as an existing one?

**Comment by thomasstorm** [comment]
> I need some more clarifications:
> 
> 1. when I create a user group, am I automatically the group admin?

Yes.

> 2. can every user create a user group?

Yes.

> 3. how do I restrict access to a usergroup to read only? Based on the output:
>    `geodb.get_access_rights(collection_name)`
>    it looks like the other members may modify the collection:

True. You currently cannot set group members to read-only, but it could be added of course. Do we have a requirement for that? 

> 
> 4. can I delete a user-group?

Same as above - not yet but could be added if there's the requirement.


> 5. in the notebook:
> 
> > Create a new group (ensure that it does not already exist)
> 
> shouldn't it be ensured on the client side, that I cannot create a user group with the same name as an existing one?

Yes, and it already is - an exception is thrown if you're trying to create a group that already exists. The wording in the notebook is not optimal.

**Comment by thomasstorm** [comment]
> > 5. in the notebook:
> > 
> > > Create a new group (ensure that it does not already exist)
> > 
> > 
> > shouldn't it be ensured on the client side, that I cannot create a user group with the same name as an existing one?
> 
> Yes, and it already is - an exception is thrown if you're trying to create a group that already exists. The wording in the notebook is not optimal.

Fixed the wording.

**Comment by AliceBalfanz** [comment]
> AB: can every user create a user group?
> TS: Yes
Can we restrict only users who have "manage" in their subscription name to create a user group? So eurodatacube-geodbmanage and manage  be the only users to be allowed to create user groups?

current subscriptions: 

> "accesslarge"
> "basic"
> "eurodatacube-geodbaccesslarge"
> "eurodatacube-geodbbasic"
> "Euro Data Cube - geoDB: Free Trial"
> "eurodatacube-geodbmanage"
> "Euro Data Cube - xcube Generator: Starter"
> "Euro Data Cube - xcube geoDB: Free"
> "manage"
> "user"

> AB: how do I restrict access to a usergroup to read only?
> TS: True. You currently cannot set group members to read-only, but it could be added of course. Do we have a requirement for that?
No, actually there is no use case for that, as we can grant read-only access already to users. So not needed.


> AB: can I delete a user-group?
> TS: Same as above - not yet but could be added if there's the requirement.
Yes, I think as a user group admin, I should be able to delete a user group again.

**Comment by thomasstorm** [comment]
> > AB: can every user create a user group?
> > TS: Yes
> 
> Can we restrict only users who have "manage" in their subscription name to create a user group? So eurodatacube-geodbmanage and manage be the only users to be allowed to create user groups?
> [...]

Yes, good idea! Will add that function.

> > AB: how do I restrict access to a usergroup to read only?
> > TS: True. You currently cannot set group members to read-only, but it could be added of course. Do we have a requirement for that?
> 
> No, actually there is no use case for that, as we can grant read-only access already to users. So not needed.
> 
> > AB: can I delete a user-group?
> > TS: Same as above - not yet but could be added if there's the requirement.
> 
> Yes, I think as a user group admin, I should be able to delete a user group again.

Ok - adding that function!

**Comment by thomasstorm** [comment]
> > AB: Yes, I think as a user group admin, I should be able to delete a user group again.
> 
> TS: Ok - adding that function!

Actually, I am not. Adding this function means considerable security issues - generally spoken, you have to be superuser to remove a user group, and allowing for superuser access over PostGREST is a delicate matter. Since I am not an expert on database security, I'd prefer to not add that function until we receive the respective external requirement.

**Comment by thomasstorm** [comment]
> > > AB: can every user create a user group?
> > > TS: Yes
> > 
> > 
> > Can we restrict only users who have "manage" in their subscription name to create a user group? So eurodatacube-geodbmanage and manage be the only users to be allowed to create user groups?
> > [...]
> 
> TS: Yes, good idea! Will add that function.

Done. Ready for re-review now!

**Review by AliceBalfanz — APPROVED** [review]
Thanks for the enhancements! works nicely :) 
Please see my inline suggestions before merging (I edited them slightly)

**Pull request merged**